web/src/features/public-api/server/RateLimitService.ts

```diff
@@ -0,0 +1,216 @@
+import type Redis from "ioredis";
+import { type z } from "zod";
+import { RateLimiterRedis, RateLimiterRes } from "rate-limiter-flexible";
+import { env } from "@/src/env.mjs";
+import {
+  type RateLimitResult,
+  type RateLimitResource,
+  type RateLimitConfig,
+  type Plan,
+} from "@langfuse/shared";
+import {
+  recordIncrement,
+  type ApiAccessScope,
+} from "@langfuse/shared/src/server";
+import { type NextApiResponse } from "next";
+
+// Business Logic
+// - rate limit strategy is based on org-id, org plan, and resources. Rate limits are appliead in buckets of minutes.
+// - rate limits are not applied for self hosters and are also not applied when Redis is not available
+// - infos for rate-limits are taken from the API access scope. Info for this scope is stored alongside API Keys in Redis for efficient access.
+// - isRateLimited returns false for self-hosters
+// - sendRestResponseIfLimited sends a 429 response with headers if the rate limit is exceeded. Return this from the route handler.
+export class RateLimitService {
+  private redis: Redis | null;
+
+  constructor(redis: Redis | null) {
+    this.redis = redis;
+  }
+
+  async rateLimitRequest(
+    scope: ApiAccessScope,
+    resource: z.infer<typeof RateLimitResource>,
+  ) {
+    // if cloud config is not present, we don't apply rate limits and just return
+    if (!env.NEXT_PUBLIC_LANGFUSE_CLOUD_REGION) {
+      return new RateLimitHelper(undefined);
+    }
+
+    if (!this.redis) {
+      console.log("Rate limiting not available without Redis");
+      return new RateLimitHelper(undefined);
+    }
+
+    return new RateLimitHelper(await this.checkRateLimit(scope, resource));
+  }
+
+  async checkRateLimit(
+    scope: ApiAccessScope,
+    resource: z.infer<typeof RateLimitResource>,
+  ) {
+    const planBasedConfig = getRateLimitConfig(scope.plan, resource);
+
+    const customConfig = scope.rateLimits?.find(
+      (config) => config.resource === resource,
+    );
+
+    const effectiveConfig = customConfig || planBasedConfig;
```

Comment:
nit" i'd suggest ot move this into getRateLimitConfig as we can then reuse that function to render rate limits in the UI (iterating over all rate limits resources, can be in trpc route) while keeping the logic of overrides in a central function